File: src/main/java/com/google/android/finsky/api/model/ContainerList.java
package com.google.android.finsky.api.model;

import com.anod.appwatcher.utils.CollectionsUtils;
import com.google.android.finsky.api.DfeUtils;
import com.google.android.finsky.protos.nano.Messages;
import com.google.android.finsky.protos.nano.Messages.DocV2;

import java.util.Arrays;
import java.util.List;

public abstract class ContainerList<T> extends PaginatedList<T, Document>
{
    private final CollectionsUtils.Predicate<Document> responseFiler;

    ContainerList(final String url, final boolean autoLoadNextPage, CollectionsUtils.Predicate<Document> responseFilter) {
        super(url, autoLoadNextPage);
        this.responseFiler = responseFilter;
    }
    
    ContainerList(final List<UrlOffsetPair> list, final int count, final boolean autoLoadNextPage, CollectionsUtils.Predicate<Document> responseFilter) {
        super(list, count, autoLoadNextPage);
        this.responseFiler = responseFilter;
    }

    @Override
    protected Document[] getItemsFromResponse(Messages.Response.ResponseWrapper wrapper) {
        Messages.Response.Payload payload = payload(wrapper);
        Messages.DocV2 doc = DfeUtils.getRootDoc(payload);

        if (doc == null) {
            return new Document[0];
        }

        Document[] docs = this.getItems(doc);
        if (responseFiler == null) {
            return docs;
        }
        List<Document> list = CollectionsUtils.INSTANCE.filter(Arrays.asList(docs), responseFiler);
        return list.toArray(new Document[0]);
    }

    @Override
    protected String getNextPageUrl(Messages.Response.ResponseWrapper wrapper) {
        Messages.Response.Payload payload = payload(wrapper);
        Messages.DocV2 doc = DfeUtils.getRootDoc(payload);

        if (doc == null) {
            return null;
        }

        final Messages.Containers.ContainerMetadata containerMetadata = doc.containerMetadata;
        String nextPageUrl = null;
        if (containerMetadata != null) {
            nextPageUrl = doc.containerMetadata.nextPageUrl;
        }
        return nextPageUrl;
    }

    private Messages.Response.Payload payload(Messages.Response.ResponseWrapper wrapper) {
        Messages.Response.Payload payload = wrapper.payload;
        if (wrapper.preFetch.length > 0
                && ((payload.searchResponse != null && payload.searchResponse.doc.length == 0)
                || (payload.listResponse != null && payload.listResponse.doc.length == 0)
        )
                ) {
            return wrapper.preFetch[0].response.payload;
        }
        return payload;
    }

    private Document[] getItems(final DocV2 docV2) {
        Document[] array;
        if (docV2 != null) {
            final int length = docV2.child.length;
            array = new Document[length];
            for (int i = 0; i < length; ++i) {
                array[i] = new Document(docV2.child[i]);
            }
        }
        else {
            array = new Document[0];
        }
        return array;
    }
}

File: src/main/java/com/google/android/finsky/api/model/ContainerList.kt
package com.google.android.finsky.api.model

import com.anod.appwatcher.utils.CollectionsUtils
import com.google.android.finsky.api.DfeUtils
import com.google.android.finsky.protos.nano.Messages

abstract class ContainerList<T> constructor(
        url: String,
        autoLoadNextPage: Boolean,
        responseFilter: CollectionsUtils.Predicate<Document>
    ) : PaginatedList<T, Document>(url, autoLoadNextPage) {

    private val responseFiler: CollectionsUtils.Predicate<Document>? = responseFilter

    override fun getItemsFromResponse(wrapper: Messages.Response.ResponseWrapper): Array<Document> {
        val payload = payload(wrapper)
        val doc = DfeUtils.getRootDoc(payload) ?: return arrayOf()

        val docs = doc.child?.map { Document(it) } ?: listOf()
        if (responseFiler == null) {
            return docs.toTypedArray()
        }
        val list = CollectionsUtils.filter(docs, responseFiler)
        return list.toTypedArray()
    }

    override fun getNextPageUrl(wrapper: Messages.Response.ResponseWrapper): String? {
        val payload = payload(wrapper)
        val doc = DfeUtils.getRootDoc(payload) ?: return null

        val containerMetadata = doc.containerMetadata
        var nextPageUrl: String? = null
        if (containerMetadata != null) {
            nextPageUrl = doc.containerMetadata.nextPageUrl
        }
        return nextPageUrl
    }

    private fun payload(wrapper: Messages.Response.ResponseWrapper): Messages.Response.Payload {
        val payload = wrapper.payload
        return if (wrapper.preFetch.isNotEmpty() && (payload.searchResponse != null && payload.searchResponse.doc.isEmpty() || payload.listResponse != null && payload.listResponse.doc.isEmpty())) {
            wrapper.preFetch[0].response.payload
        } else payload
    }

}
